sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala

```diff
@@ -0,0 +1,356 @@
+      // can cause error like NPE.
+      (expr.find(_.isInstanceOf[PlanExpression[_]]).isDefined && TaskContext.get != null)
+
+    if (!skip && !addExprToMap(expr, map)) {
+      childrenToRecurse(expr).foreach(addExprTree(_, map))
+      commonChildrenToRecurse(expr).filter(_.nonEmpty).foreach(addCommonExprs(_, map))
+    }
+  }
+
+  /**
+   * Returns the state of the given expression in the `equivalenceMap`. Returns None if there is no
+   * equivalent expressions.
+   */
+  def getExprState(e: Expression): Option[GpuExpressionStats] = {
+    equivalenceMap.get(GpuExpressionEquals(e))
+  }
+
+  // Exposed for testing.
+  private[sql] def getAllExprStates(count: Int = 0): Seq[GpuExpressionStats] = {
+    equivalenceMap.values.filter(_.useCount > count).toSeq.sortBy(_.height)
+  }
+
+  /**
+   * Returns a sequence of expressions that more than one equivalent expressions.
+   */
+  def getCommonSubexpressions: Seq[Expression] = {
+    getAllExprStates(1).map(_.expr)
+  }
+
+  /**
+   * Returns the state of the data structure as a string. If `all` is false, skips sets of
+   * equivalent expressions with cardinality 1.
+   */
+  def debugString(all: Boolean = false): String = {
+    val sb = new java.lang.StringBuilder()
+    sb.append("GPU Equivalent expressions:\n")
+    equivalenceMap.values.filter(stats => all || stats.useCount > 1).foreach { stats =>
+      sb.append("  ").append(s"${stats.expr}: useCount = ${stats.useCount}").append('\n')
+    }
+    sb.toString()
+  }
+}
+
+object GpuEquivalentExpressions extends Logging {
+  def logExprTree(expr : Expression): Unit = {
+    logWarning("nodeName: " + expr.nodeName + "\n" + expr.treeString)
+  }
+
+  def logExprs(title: String, exprs: Seq[Expression]): Unit = {
+    if (!exprs.isEmpty) {
+      logWarning("Start: " + title)
+      exprs.foreach(logExprTree)
+      logWarning("  End: " + title)
+    }
+  }
+
+  def logExprTiers(title: String, tiers: Seq[Seq[Expression]]): Unit = {
+    tiers.foreach(logExprs(title, _))
+  }
+
+  /**
+   * Recursively replaces expression with its proxy expression in `substitutionMap`.
+   */
+  def replaceWithCommonRef(
+      expr: Expression,
+      substitutionMap: Map[Expression, Expression]): Expression = {
+    expr match {
+      case e : AttributeReference => expr
+      case _ =>
+        substitutionMap.get(expr) match {
+          case Some(e) => e
+          case None => expr.mapChildren(replaceWithCommonRef(_, substitutionMap))
+        }
+    }
+  }
+
+  // Given a set of expressions, recursively extract all of the common expressions
+  def getExprTiers(expressions : Seq[Expression]): Seq[Seq[Expression]] = {
+    def recurse(exprs: Seq[Expression], startIdx: Int): Seq[Seq[Expression]] = {
+      val equivalentExpressions = new GpuEquivalentExpressions
+
+      // Filter out some expressions that we know are problematic
+      // This won't be necessary if we replace EquivalentExpressions
```

Comment:
Didn't we replace EquivalentExpressions???

Reply:
Yes. And I removed that filter as well.  I will remove the comment.

Reply:
Comment removed.